### ISSUE TITLE
[syzkaller] memory leak in tcp_md5_do_add

### ISSUE THREAD

**Issue by cpaasch** [multipath-tcp/mptcp_net-next#174]
```
sendmsg$NLBL_MGMT_C_ADDDEF(r1, &(0x7f0000000180)={0x0, 0x0, &(0x7f0000000140)={&(0x7f0000000240)=ANY=[@ANYBLOB="14000000", @ANYRES16=0x0, @ANYBLOB="009375253b32343b5d1a0408000078644760e453b4c883b8dd204d372e6514c62b0e277df8295dda6f0827c36258f3cd65d3ca0ad6790cc651233a01000000093e3019219bbd33315cf7dc8ffc8e58e02cb2d0aa94910ddcdd39de48317adabe9509325cad3dcc0981cc8ed5a2de606227a16c6200"/127], 0x14}}, 0x88c1)
BUG: memory leak
unreferenced object 0xffff888108ea8fc0 (size 32):
  comm "syz-executor.0", pid 14707, jiffies 4295891566 (age 24.585s)
  hex dump (first 32 bytes):
    00 08 6a 23 80 88 ff ff 76 32 00 00 7f 00 60 aa  ..j#....v2....`.
    c0 83 ea 08 81 88 ff ff b0 90 7b 81 ff ff ff ff  ..........{.....
  backtrace:
    [<000000004bb91791>] kmalloc include/linux/slab.h:554 [inline]
    [<000000004bb91791>] tcp_md5_do_add+0x375/0x530 net/ipv4/tcp_ipv4.c:1163
    [<000000001c9ae5bc>] tcp_v4_parse_md5_keys+0x1ef/0x440 net/ipv4/tcp_ipv4.c:1275
    [<00000000c643c9d8>] do_tcp_setsockopt net/ipv4/tcp.c:3554 [inline]
    [<00000000c643c9d8>] tcp_setsockopt+0x845/0x25f0 net/ipv4/tcp.c:3638
    [<0000000025c26205>] mptcp_setsockopt+0x51d/0x650 net/mptcp/protocol.c:2901
    [<000000009614680e>] __sys_setsockopt+0x14f/0x360 net/socket.c:2115
    [<00000000ced56819>] __do_sys_setsockopt net/socket.c:2126 [inline]
    [<00000000ced56819>] __se_sys_setsockopt net/socket.c:2123 [inline]
    [<00000000ced56819>] __x64_sys_setsockopt+0xb9/0x150 net/socket.c:2123
    [<00000000f5d8be7e>] do_syscall_64+0x33/0x40 arch/x86/entry/common.c:46
    [<00000000a012b8c7>] entry_SYSCALL_64_after_hwframe+0x44/0xae

BUG: memory leak
unreferenced object 0xffff8880236a0800 (size 192):
  comm "syz-executor.0", pid 14707, jiffies 4295891566 (age 24.585s)
  hex dump (first 32 bytes):
    00 00 00 00 00 00 00 00 c0 8f ea 08 81 88 ff ff  ................
    20 02 20 00 ac 14 14 bb 00 00 00 00 00 00 00 00   . .............
  backtrace:
    [<00000000ec34ad94>] kmalloc include/linux/slab.h:559 [inline]
    [<00000000ec34ad94>] sock_kmalloc net/core/sock.c:2263 [inline]
    [<00000000ec34ad94>] sock_kmalloc+0xb5/0x100 net/core/sock.c:2254
    [<000000007ada111b>] tcp_md5_do_add+0x107/0x530 net/ipv4/tcp_ipv4.c:1172
    [<000000001c9ae5bc>] tcp_v4_parse_md5_keys+0x1ef/0x440 net/ipv4/tcp_ipv4.c:1275
    [<00000000c643c9d8>] do_tcp_setsockopt net/ipv4/tcp.c:3554 [inline]
    [<00000000c643c9d8>] tcp_setsockopt+0x845/0x25f0 net/ipv4/tcp.c:3638
    [<0000000025c26205>] mptcp_setsockopt+0x51d/0x650 net/mptcp/protocol.c:2901
    [<000000009614680e>] __sys_setsockopt+0x14f/0x360 net/socket.c:2115
    [<00000000ced56819>] __do_sys_setsockopt net/socket.c:2126 [inline]
    [<00000000ced56819>] __se_sys_setsockopt net/socket.c:2123 [inline]
    [<00000000ced56819>] __x64_sys_setsockopt+0xb9/0x150 net/socket.c:2123
    [<00000000f5d8be7e>] do_syscall_64+0x33/0x40 arch/x86/entry/common.c:46
    [<00000000a012b8c7>] entry_SYSCALL_64_after_hwframe+0x44/0xae

BUG: leak checking failed
```

HEAD:
4dc9bac42277 ("DO-NOT-MERGE: mptcp: enabled by default")  (HEAD, tag: export/20210226T185626, mptcp_net-next/export) (2 weeks ago) <Matthieu Baerts>
271fd8bbbb68 ("DO-NOT-MERGE: mptcp: add GitHub Actions")  (2 weeks ago) <Matthieu Baerts>
085c6081951b ("DO-NOT-MERGE: mptcp: use kmalloc on kasan build")  (2 weeks ago) <Paolo Abeni>
a36b8de12bfe ("selftests: mptcp: dump more info on mpjoin errors")  (2 weeks ago) <Matthieu Baerts>
41ac43dba21a ("selftests: mptcp: init nstat history")  (2 weeks ago) <Matthieu Baerts>
6a7d55594d5c ("selftests: mptcp: launch mptcp_connect with timeout")  (2 weeks ago) <Matthieu Baerts>
9325d7ed5a01 ("mptcp: add mptcp reset option support")  (2 weeks ago) <Florian Westphal>
5f8de0ccd9ee ("mptcp: remove unneeded check on first subflow")  (2 weeks ago) <Paolo Abeni>
1f02e44c0143 ("mptcp: add active MPC mibs")  (2 weeks ago) <Paolo Abeni>
c43bff3b4e67 ("mptcp: add mib for token creation fallback")  (2 weeks ago) <Paolo Abeni>
2d103a9a604e ("selftests: mptcp: signal addresses testcases")  (2 weeks ago) <Geliang Tang>
a146d882ca38 ("mptcp: move to next addr when subflow creation fail")  (2 weeks ago) <Geliang Tang>
7ff83ceef7dd ("mptcp: export lookup_anno_list_by_saddr")  (2 weeks ago) <Geliang Tang>
c885ea369a1a ("selftests: mptcp: add testcases for removing addrs")  (2 weeks ago) <Geliang Tang>
318e0559e0bc ("selftests: mptcp: set addr id for removing testcases")  (2 weeks ago) <Geliang Tang>
524fba068122 ("selftests: mptcp: add invert argument for chk_rm_nr")  (2 weeks ago) <Geliang Tang>
e1398867c1aa ("mptcp: remove a list of addrs when flushing")  (2 weeks ago) <Geliang Tang>
edb9620f83e6 ("mptcp: remove multi addresses and subflows in PM")  (2 weeks ago) <Geliang Tang>
aca6f0dd3914 ("mptcp: remove multi subflows in PM")  (2 weeks ago) <Geliang Tang>
71a9f0c3adfc ("mptcp: remove multi addresses in PM")  (2 weeks ago) <Geliang Tang>
daf1fc16c8f8 ("mptcp: add rm_list_rx in mptcp_pm_data")  (2 weeks ago) <Geliang Tang>
e6c556c40432 ("mptcp: add rm_list in mptcp_options_received")  (2 weeks ago) <Geliang Tang>
64ef994b0a5f ("mptcp: add rm_list_tx in mptcp_pm_data")  (2 weeks ago) <Geliang Tang>
8d960b024487 ("mptcp: add rm_list in mptcp_out_options")  (2 weeks ago) <Geliang Tang>
a398ebc3af80 ("selftests: mptcp: timeout testcases for multi addresses")  (2 weeks ago) <Geliang Tang>
ce3735dc6ff9 ("selftests: mptcp: add cfg_do_w for cfg_remove")  (2 weeks ago) <Geliang Tang>
d13974555483 ("mptcp: move to next addr when timeout")  (2 weeks ago) <Geliang Tang>
bf4629e714a7 ("mptcp: drop unused subflow in mptcp_pm_subflow_established")  (2 weeks ago) <Geliang Tang>
56e9734535b6 ("mptcp: skip connecting the connected address")  (2 weeks ago) <Geliang Tang>
f3dd90f9211c ("mptcp: drop argument port from mptcp_pm_announce_addr")  (2 weeks ago) <Geliang Tang>
507edafb61ae ("mptcp: free resources when the port number is mismatched")  (2 weeks ago) <Geliang Tang>
1884902470ae ("mptcp: fix missing wakeup")  (2 weeks ago) <Paolo Abeni>
c530cf9a4f22 ("mptcp: reset 'first' and ack_hint on subflow close")  (2 weeks ago) <Florian Westphal>
faa6e7b09286 ("mptcp: fix DATA_FIN generation on early shutdown")  (2 weeks ago) <Paolo Abeni>
bc332c878c91 ("mptcp: clean-up the rtx path")  (2 weeks ago) <Paolo Abeni>
744fcac06ce5 ("mptcp: fix race in release_cb")  (2 weeks ago) <Paolo Abeni>
888b67341995 ("mptcp: factor out __mptcp_retrans helper()")  (2 weeks ago) <Paolo Abeni>
9b17fa5ec0f4 ("mptcp: dispose initial struct socket when its subflow is closed")  (2 weeks ago) <Florian Westphal>
d93f13f44883 ("bpf:selftests: add bpf_mptcp_sock() verifier tests")  (2 weeks ago) <Nicolas Rybowski>
fd936530445b ("bpf:selftests: add MPTCP test base")  (2 weeks ago) <Nicolas Rybowski>
6b4221484f0d ("bpf: add 'bpf_mptcp_sock' structure and helper")  (2 weeks ago) <Nicolas Rybowski>
e3071ba9a0ee ("bpf: expose is_mptcp flag to bpf_tcp_sock")  (2 weeks ago) <Nicolas Rybowski>
30863463de43 ("linux: handle MPTCP consistently with TCP")  (2 weeks ago) <Paolo Abeni>
6e6c2efd5ae9 ("mptcp: fix memory accounting on allocation error")  (2 weeks ago) <Paolo Abeni>
b5b1a24df0fa ("mptcp: do not wakeup listener for MPJ subflows")  (2 weeks ago) <Paolo Abeni>
644953273b1d ("mptcp: put subflow sock on connect error")  (2 weeks ago) <Florian Westphal>
f85f77200a6d ("mptcp: fix DATA_FIN processing for orphaned sockets")  (2 weeks ago) <Paolo Abeni>
32fcc880e0a9 ("mptcp: provide subflow aware release function")  (2 weeks ago) <Florian Westphal>
fc43ad8c2616 ("mptcp: reset last_snd on subflow close")  (2 weeks ago) <Florian Westphal>
d310ec03a34e ("Merge tag 'perf-core-2021-02-17' of git://git.kernel.org/pub/scm/linux/kernel/git/tip/tip")  (mptcp_net-next/net-next) (3 weeks ago) <Linus Torvalds>

CONFIG-file: 
[CONFIG.txt](https://github.com/multipath-tcp/mptcp_net-next/files/6142546/CONFIG.txt)

**Comment by matttbe** [comment]
Hi Christoph,

May you retry with the latest export branch please? This is probably fixed by the new sockopt white/allowed list added by Paolo to fix #170 
→ - ce35fc3cba0b: mptcp: only admit explicitly supported sockopt

We should not accept MD5 sockopt.

**Comment by matttbe** [comment]
Christoph didn't see this one with the export branch → Done
